### ISSUE TITLE
[Bug] Notifications.   (android issue???)

### ISSUE THREAD

**Issue by boydkelly** [ankidroid/Anki-Android#8114]
When I have notifications enabled, i get notified over and over and over and over and over and over and over and over...

If not 5 seconds then 5 min after I dismiss.

Is this an android problem?   I wouldn't mind getting a notification once (or heck twice) a day.   But 5 seconds after I dismiss the anki notification it comes back.   Why is this happeing?

**Comment by mikehardy** [comment]
Mostly because our notifications sub-system needs a *lot* of work, and it hasn't been a priority. I feel your pain but also just swipe away the notifications as I haven't had time personally to implement a fix :shrug:

**Comment by david-allison-1** [comment]
It's to do with the widget, Best to disable the notifications in the settings. Mirroring mike, we need someone who cares about the notifications to spend a few days fixing them, but it hasn't hit my priority yet as I don't use them.

**Comment by moqi-Git** [comment]
I want to try fixing this but I can't reproduce on my device. Which kind of notification bothered you?

**Comment by boydkelly** [comment]
Hi,
I was probably in a grumpy mood the morning I posted that, as this is nothing seriously affecting Anki.   But I have under settings general notifications.  Notify when more than 75 cards due.   Probably because I have 4 decks and several filtered decks, I get 3 or 4 notifications when I look at my phone in the morning.   But the issue is if I dismiss the notifications, because I just can't study, they will come back again, sometimes within seconds, and I have to keep dismissing them.   Hey I was actually finding that kind of funny at first that Anki was so 'determined' to get me on the right track!   But then after a while with constant notifications all day it becomes wearysome.   So they are off now.   No biggie.   If you can make notifications come on only once in a 24 hr period that would be better IMO.

**Comment by mikehardy** [comment]
I reproduce the same behavior, and think I even wrote the code that did it! Because, awesome :sweat_smile: - hey, more notifications is better than 0? But in this case only slightly

**Comment by 2tanayk** [comment]
> Mostly because our notifications sub-system needs a _lot_ of work, and it hasn't been a priority. I feel your pain but also just swipe away the notifications as I haven't had time personally to implement a fix 🤷

will do the work on the notifications thing, let me reproduce the issue and I will get started then 👍

**Comment by 2tanayk** [comment]
Thanks for assigning me this issue @david-allison-1 sir!
It would really be great if you provide me with a bit of guidance for this issue as I do not know much about the `notifications sub-system` (although I will search and start studying the code as soon as possible). 
Just brief me about any key aspects of how it works right now, Thanks in advance!

**Comment by mikehardy** [comment]
#4944

**Comment by 2tanayk** [comment]
@mikehardy sir first of all thanks for linking that discussion to me,but that discussion is 2 years old and as someone who is new to this organisation (I might be asking too much) I would really like to get a gist of what exactly needs to be done, like in points, would really help. That being said I've already started studying the code and using the existing resources to understand the current system in depth

**Comment by david-allison-1** [comment]
https://developer.android.com/guide/topics/ui/notifiers/notifications

https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java

https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/NotificationChannels.java

and related history

**Comment by Arthur-Milchior** [comment]
Do you get one notification by deck or do you get the notification for the same deck everytime? Because one problem is that there are notifications for each top level deck potentially. I don't like it but was told to let it this way. 
I can imagine that android limit notification to a small number and queue the next notification until the first one are seen.

**Comment by mikehardy** [comment]
Android I think has a limit of 50 notifications at once, and some phones (Samsung I think?) limit to 500 pending intents (think "alarms") set at once.

I think the first thing to solve is that if notifications are swiped away for the day, they actually stay away until the next day

**Comment by 2tanayk** [comment]
Working on this

**Comment by haran2248** [comment]
Can I try taking up this issue as well?

**Comment by 2tanayk** [comment]
> Can I try taking up this issue as well?If it is ok with @2tanayk

I don't think two people should work at the same time

**Comment by prateek-singh-3212** [comment]
Hi @2tanayk,
If you have any need, I would love to help you with this issue.

**Comment by 2tanayk** [comment]
> Hi @2tanayk,
> If you have any need, I would love to help you with this issue.

Sure thanks!

**Comment by 2tanayk** [comment]
I am working on it,was a bit busy(because of college 😬),will get it done soon :)

**Comment by david-allison-1** [comment]
@2tanayk Life comes first, take your time, grades are important.

**Comment by 2tanayk** [comment]
if some1 else wants they can take this issue as I have not been able to spend ample time at this ,really sorry for any inconvenience caused :(,but I'll definitely attend to it if nobody does

**Comment by david-allison-1** [comment]
Don't worry! Life comes first. Give me a ping when you have time

**Comment by eskoONE** [comment]
> Can I try taking up this issue as well?If it is ok with @2tanayk



> Hi @2tanayk,
> If you have any need, I would love to help you with this issue.

Do you guys want to have another look at this issue? The guy that wanted to work on this didn't have the time to do so and left it open for others. Since nobody seems to be working on this anymore, I wanted to make sure you guys have seen this, since you showed interest on working on this issue. 

I think ankidroid is in a dire need of a notification system overhaul, since its practically not very useful and difficult to use (perhaps not really difficult, more like confusing. Its weird to have a widget on the home screen to have notifications to work at all).

Would appreacute of one of you guys or both of you could figure out a better way of implementing notifications for ankidroid.

Thanks!

**Comment by eskoONE** [comment]
I meant to ping @haran2248 and @prateek-singh-3212, please read the comment above.

Sry about the double post.

**Comment by prateek-singh-3212** [comment]
@eskoONE  Thanks for reminding me about this issue. Currently their are 3 issues of ankidroid in my todo. But I will take this one because I want to brush up my notification concepts.

**Comment by david-allison-1** [comment]
@eskoONE Would you be interested in taking this on? 

I feel the design on this one is going to be more work than the coding.

**Comment by eskoONE** [comment]
> @eskoONE Thanks for reminding me about this issue. Currently their are 3 issues of ankidroid in my todo. But I will take this one because I want to brush up my notification concepts.

@prateek-singh-3212 Great, thank you for taking this one on you.

@david-allison-1 unfortunately my coding skills are very limited and I don't know any java at all. If there is some mundane python scripting to be done, perhaps I could be of use.

**Comment by prateek-singh-3212** [comment]
Hello,
I am planning to improve the notification system this month. So, if anyone is having an idea or plan that when notification should trigger then please share. Otherwise I will make the plan. 

I would like to purpose the idea that should we implement the button in the notification `set reminder` from which they were able to set reminder in their calendar.  This helps the user in case when he/she is not able to review at moment. 
NOTE: This button is only for REMINDERS channel.

**Comment by david-allison-1** [comment]
> Hello,
> I am planning to improve the notification system this month. So, if anyone is having an idea or plan that when notification should trigger then please share. Otherwise I will make the plan.
> 
> I would like to purpose the idea that should we implement the button in the notification `set reminder` from which they were able to set reminder in their calendar. This helps the user in case when he/she is not able to review at moment.
> NOTE: This button is only for REMINDERS channel.

It'd be good to have a design before we start an implementation here.

Mike's added a potential design in another issue tagged with "Advanced Reminders"

**Comment by eskoONE** [comment]
one thing that is annoying about the current notification system is that you need a widget placed in home screen to get notifications at all. would be nice if we could have it working without it, and if its necessary, make it 1x1 instead of whatever that it is rn.

perhaps being able to set the exact minimum amount of reviews before getting notified would be handy. i had use cases where i would have less than 10 reviews that i was waiting for and couldnt get notified unless i set it to any review. that would trigger the notification all the time after the first one though. pretty pointless imo.

one could make use of led for notifications. perhaps setting different sounds/colors for different decks as well.

**Comment by prateek-singh-3212** [comment]
**USE OF NOTIFICATION**

- To Send Deck Reminder.
- Cards Due
- Synchronization

**ALGORITHMS FOR NOTIFICATION**

1. **DECK REMINDER** :- We can setup the new ` NOTIFICATION` setting this setting screen looks like this [(FIGMA LINK)](https://www.figma.com/file/Mf36OjE4S1eNGKVqWtIeRJ/Untitled?node-id=0%3A1). We can store the timing in DATABASE and trigger the Notification at that perticular time using ALARM MANAGER. 

2. **CARD DUE** :-  General card DUE Notification can be send by  setisfying these conditions:-
      

- When DUE card is above the limit set by user.
- Time specified by user (BY DEFAULT notification can be send early morning accoring to time zone of user)

3. **Synchronization** :- In material Guidline it is not recommended to show the notification to user which does not require user attention ( [this](https://material.io/design/platform-guidance/android-notifications.html#usage) When not to use a notification -> POINT 5) but if we want to use notification then we can create A function and whenever we want to show sync notification we can call that and whenever sync is complete we have to manually close this notification.

THIS IS THE PLAN FROM MY SIDE 😃

**Comment by david-allison-1** [comment]
You can't modify notification channels after they've been created, that will complicate the design.

I think taking on "Sync" notifications in this would be too much work, let's focus on one component at a time, and make it good.

I think the first thing that we should tackle is investigating the requirement for the widget and whether we can remove this. I feel it's almost certainly the case and will produce the most value to our users.

**Comment by mikehardy** [comment]
The requirement for the widget to exist may *certainly* be removed with proper code
I literally maintain a notifications package (https://notifee.app) and there is no requirement on Android's part to have a widget running in order to get notifications. It's just something we're doing oddly in AnkiDroid and I never had time to root it out

**Comment by prateek-singh-3212** [comment]
Hello, 
I planned the notification design which **Dose'nt** Require the Widget.
Here is the short proposal. Please comment the if their is something I missed or did wrong. 

@david-allison-1 Regarding the UI thing I will update it and notify here. I think the Removal of widget from notification and Adding new deck notification UI is different task.

**Comment by mikehardy** [comment]
@prateek-singh-3212 - can I ask you to keep all discussion here? I read the notion link and I see no reason why simple markdown in a comment here can't capture it, that way we have all the discussion in one place that won't suffer link rot over the next 10 years...

**Comment by prateek-singh-3212** [comment]
## What we are doing wrong?

The main issue in making notification functional is that whenever the widget status updates notification triggers in POST Execute ([this](https://github.com/ankidroid/Anki-Android/blob/09893d5cd41e5f49eeb2bb1fa59e00e02bdb75df/AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.java#L114)).

 Whenever these (improper) condition occurs card due  displays notification :-

1. Settings opens.
2. Deck is Pressed (because widgets updates the deck count)
3. Deck Review goes to OnPause (because  widgets updates the deck count).

## Possible Solution

We can make the new repository which holds the meta Data (or Meta information) of all decks like Total deck cards, Total New cards to Review, Total Learning cards etc in sqlite Database and whenever the data is required by widget , Notification or any other activity, Service.

This discussion is documented in this PR ([this](https://github.com/ankidroid/Anki-Android/pull/5642)). 

### Flow Chart of Above Explained.

![untitled](https://user-images.githubusercontent.com/76490368/128639673-1a948cea-980d-4bfa-a6d2-7da5c5670d9f.png)


### Working

Repository class works asyncronously and fetches the metadata from collection when app Starts and updates the table and after the app goes to onPause or onDestroy.

**Repository Table Structure: (***Similar to widgetStatus table but I think we are not using this (?)***)**

1. **did** :- Deck Id
2. **deckName** :- Deck Name
3. **newCards :-** Total new cards of deck
4. **lrnCards** :- Total learning cards of deck
5. **dueCards :-**  Total due cards of deck.
6. **eta :-** Expected time to complete.

Whenever we need the data we call the repositry function which fetches fresh data from Database.

### Limitations

1. We are not able to call this in main thread (UI Thread)
2. We are using the broadcast receiver which runs on the UI Thread. So Database Operation is allowed. (I don't know that can we run async task from broadcast receiver)
    - **Solution:** We can use the WorkManager. It is not accurate in timing. ([reference](https://martian.ventures/mantra/insights/schedule-local-notifications-on-android/))

**Comment by mikehardy** [comment]
With regard to limitations: WorkManager should be sufficient for our uses, this is not an alarm clock application that needs precision.

With regard to design, what if: 
- we just stored which decks users wanted notifications for, and what time to display the notification
- we set a workmanager task to generate the notifications at that time
- then at all the other times (onPause etc) we had an async task that 1) scanned for any existing notifications that had not been dismissed and 2) for any non-dismissed existing notifications, updated their accounts quietly just so the notification was up to date

Really really simple, but working, in other words. No need for database stuff, not a lot of code ?

The hard part of the whole notification thing - I thought - was user presentation. How to let users select exactly what they wanted notifications for or not

**Comment by david-allison-1** [comment]
I'm personally moving away from the concept of storing preferences in a fixed database schema, there's a lot of overhead from schema versioning, it makes it harder to inspect the data, SQL is verbose and prone to mistakes and there's little benefit (maybe speed?)

Echoing Mike: screen designs are the most useful, after that we can discuss data storage.

**Comment by prateek-singh-3212** [comment]
> The UI doesn't seem to allow for an "any/all decks" notification.

@david-allison-1 Updated UI, Hope this solves the above problem. 
Removed Vibrate and Blink Light feature. We can set by default vibrate to true and blink light to white.

Figma Link:- https://www.figma.com/file/Mf36OjE4S1eNGKVqWtIeRJ/Untitled?node-id=13%3A95
![Screenshot 2021-08-10 at 12-38-17 Figma](https://user-images.githubusercontent.com/76490368/128824136-d3f884cf-5025-48cb-9f10-442c9ad812c6.png)


**I Concluded Till Now**
We have to do 2 task.
 1. We have to remove widget from notification 
      - **solution** is proposed in [this](https://github.com/ankidroid/Anki-Android/issues/8114#issuecomment-894826018) Comment. 
      - **Issue** is with [this](https://github.com/ankidroid/Anki-Android/blob/39a226842004e9f329304bd227fee9bace7d5f37/AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java#L53) line and cases where we used this. 

 2. We have to add the separate  screen so that user can manage all notification at one place.
     - **Solution** is proposed in this comment (above ⬆️ ).
     - **Issue** we are not showing deck notification for specific time which user selects.

**Comment by mikehardy** [comment]
This looks pretty cool (the UI mockup) and it seems you are on the right track... :steam_locomotive:

**Comment by prateek-singh-3212** [comment]
Done Some changes in layout as suggested by @david-allison-1 

Figma Link:- https://www.figma.com/file/Mf36OjE4S1eNGKVqWtIeRJ/Untitled?node-id=13%3A95

![Screenshot 2021-08-11 at 23-34-39 Figma](https://user-images.githubusercontent.com/76490368/129080506-148a5c9c-fde0-4031-8739-b8c1fd0e0978.png)

**Regarding Storing Time**
Their are 3 types of storage which satisfies our needs:-
1. **SQLite**:- Most Promising way to store data. 
   - **Benifits**
     1. Fastest
     2. Able to Query Data
   - **Demerits**
     1. Requires new Thread other than UI Thread
     2. Lot's of boilerplate code
       - **Solution**
          -> We can use room which removes lots of boilerplate code and makes easier to unit test the code. I think we can use this  because we are creating seprate table which stores metadata so in future we can fetch data from rust database and store in sqlite table (I am completely unaware of rust database which we are going to implement in future, So feel free to    counter this solution )                   
2. **Shared Preferences** :- We can store in this but we need plan proper structure to use this.
     **Benifits**
        1. Able to run on Main Thread
        2. Eaiser way to store
     **Demerits**
        1. Not made to store this type of data (Decks details)
        2. Slower than sqlite

**Comment by mikehardy** [comment]
You forget the ability to store things in the deck configuration so that notifications will sync across devices...that may actually be the best

I'd like to avoid local storage if possible, especially local sql storage

I question your assertion that preferences is slower than sqlite
I want to avoid a dependency on room if possible, it is not a light dependency and will have it's own failure modes I'm sure...

**Comment by prateek-singh-3212** [comment]
> You forget the ability to store things in the deck configuration so that notifications will sync across devices...that may actually be the best

Yaa I forgot about deck Config. I think we can use `mod` and can do all the calculation of time relative to `mod` and show notifications.
But I don't know that can we edit the json model to store the deck reminder as `notification_time`. This might cause the issue with web or Anki Desktop (?). 

So, If we were able to solve problem then I think we are good to go (... 🚀) and  I will start working on this in two seprate PR's

**Comment by eskoONE** [comment]
@prateek-singh-3212 any news on this?

**Comment by prateek-singh-3212** [comment]
Working on it!!
